Driverless AI/scoring-pipeline-deployment-in-c++-runtime/scoring-pipeline-deployment-in-c++-runtime.md

```diff
@@ -0,0 +1,413 @@
+m.uuid
+```
+
+4. We can also retrieve a list of missing values, feature names, feature types, output names and output types:
+
+```
+# retrieve a list of missing values
+m.missing_values
+# retrieve the feature names
+m.feature_names
+# retrieve the feature types
+m.feature_types
+
+# retrieve the output names
+m.output_names
+# retrieve the output types
+m.output_types
+```
+
+5. Now will import the Python datatable package, load the Hydraulic System example csv data into the datatable, set the table to ignore strings that equal the missing values and display the table:
+
+```
+# import the datatable module
+import datatable as dt
+
+# parse the example.csv file
+pydt = dt.fread(homePath + "/dai-mojo-cpp/mojo-pipeline/example.csv", na_strings=m.missing_values, header=True, separator=',')
+
+pydt
+```
+
+6. We can also display the table column types:
+
+```
+# retrieve the column types
+pydt.stypes
+```
+
+7. We will use our MOJO scoring pipeline to predict the Hydraulic System’s cooling condition for each row within the table:
+
+```
+# make predictions on the example.csv file
+res = m.predict(pydt)
+
+# retrieve the predictions
+res
+```
+
+![batch-scoring-via-run-py-program-2](assets/batch-scoring-via-run-py-program-2.jpg)
+
+This classification output is the batch scoring done for our Hydraulic System cooling condition. You should receive classification probabilities for cool_cond_y.3, cool_cond_y.20, cool_cond_y.100. The 3 means the Hydraulic cooler is close to operating at total failure, 20 means it is operating at reduced efficiency and 100 means it is operating at full efficiency.
+
+8. There is some more data we can retrieve from our res predictions, which include the prediction column names, column types:
+
+
+```
+# retrieve the prediction column names
+res.names
+
+# retrieve the prediction column types
+res.stypes
+```
+
+9. We can also convert the datatable results to other data structures, such as pandas, numpy and list:
+
+```
+# need pandas
+res.to_pandas()
+
+# need numpy  
+res.to_numpy()
+
+res.to_list()
+```
+
+You just learned how to execute the MOJO scoring pipeline to do batch scoring for the Hydraulic System cooling condition using the Python wrapper in the C++ Runtime.
+
+## Task 4: Challenge
+
+### Execute Scoring Pipeline for a New Dataset
+
+There are various challenges one could do, you could do something that helps you in your daily life or job. Maybe there is a dataset you are working with, you could reproduce the steps we did above, but for your dataset, build a new experiment and execute your MOJO scoring pipeline to do batch scoring or interactive scoring.
+
+### Embed Scoring Pipeline into Existing Program
+
+Another challenge could be to use the existing MOJO scoring pipeline we executed and instead of using the examples we shared above, integrate the scoring pipeline into an existing Python or R program.
+
+## Next Steps 
+
+- Tutorial 3A: Scoring Pipeline Deployment in Java Runtime
```

Comment:
Check the number in the tutorials. I was under the impression all the deployment tutorials were going to start with Tutorial 4...